### PR TITLE
Listen on either stack for websocket connections

### PR DESCRIPTION
**Pull request by ThetaSinner**
### Summary

Apparently dual-stack for `localhost` just isn't a thing. The translation doesn't happen within machines, it is supposed to happen between machines so that machines running different IP versions can communicate. If you want to listen on either interface on localhost then you have to do that yourself.

### TODO:
- [ ] CHANGELOGs updated with appropriate info

### CODE REVIEW

File: crates/holochain_websocket/src/lib.rs
Comment (ThetaSinner) [inline crates/holochain_websocket/src/lib.rs:692]: `accept` is documented to be cancel safe so that it is acceptable to use it within a `select!`

File: crates/holochain/src/conductor/conductor.rs
Comment (ThetaSinner) [inline crates/holochain/src/conductor/conductor.rs:424]: Prefer not to do unchecked array indexing but it doesn't make sense to have a listener with no addresses so I think it's reasonable in this case

**Comment by ThetaSinner** [comment]
Unfortunately this will clash with the changes on #3622... I think it probably makes sense to get this in first and backported so that we can unblock the components team (well, just Launcher actually I think)

File: crates/holochain_websocket/src/lib.rs
Comment (neonphog) [inline crates/holochain_websocket/src/lib.rs:762]: I feel like we're going to need slightly different logic here if port is zero and not.

- non-zero: this seems incorrect, as if the ports are different, the v4 port will be overridden here, right?
- zero: there is a small chance the ephemeral port bound in v6 here has already been bound by some other process on the v4 stack, and so the v4 bind will fail. Should we have a small retry loop here to get around this?
- mixed: what do we do if one of the stacks is zero, but the other is specified?

File: crates/holochain_websocket/src/lib.rs
Comment (neonphog) [inline crates/holochain_websocket/src/lib.rs:755]: you could have the function signature take [SocketAddrV4](https://doc.rust-lang.org/std/net/struct.SocketAddrV4.html) and [SocketAddrV6](https://doc.rust-lang.org/std/net/struct.SocketAddrV6.html) instead.

File: crates/holochain_websocket/src/lib.rs
Comment (neonphog) [inline crates/holochain_websocket/src/lib.rs:741]: It is possible to configure a system with no v4 loopback device, and likewise with no v6 loopback device. In the case where only one bind is successful, do we still want this function to work, just returning only a single result in the local_addrs vec?

File: crates/holochain_websocket/src/lib.rs
Comment (neonphog) [reply crates/holochain_websocket/src/lib.rs:762]: Maybe we don't have a use-case for the ports to be different, in which case maybe we should assert they are the same. Then we just have the question of the retry loop for the ephemeral version.

**Review by neonphog — CHANGES_REQUESTED** [review]
All this looks awesome! I just think we need to sort out some edge cases before merging.

File: crates/holochain_websocket/src/lib.rs
Comment (jost-s) [inline crates/holochain_websocket/src/lib.rs:743]: If it's called `allowed_origins` this is clearer to me
```suggestion
            Error::other("WebsocketListener requires allowed origins to be set in the config")
```

File: crates/holochain_websocket/src/lib.rs
Comment (ThetaSinner) [reply crates/holochain_websocket/src/lib.rs:762]: The ports should not be different. I've added a check for this and updated the documentation for the function.

You are right that we should retry if the same port isn't available for IPv4 and IPv6. I've added a retry loop that will give us a few attempts at picking a port. That's better than the user having to restart Holochain they find a port that works!

File: crates/holochain_websocket/src/lib.rs
Comment (ThetaSinner) [reply crates/holochain_websocket/src/lib.rs:741]: Yes, you're right that is possible and we may as well handle it cleanly. I don't know how I can test this on CI but I have updated the implementation and run it with IPv6 disabled on my machine and it works nicely.

I didn't use a `vec` because the implementation gets fairly grim if you need a dynamic `accept`. The `select!` is much cleaner and easier. What I can do though is fall back to the default constructor which just binds to a single address in the case that an interface is disabled. This seems like a clean solution to me :)

File: crates/holochain_websocket/src/lib.rs
Comment (ThetaSinner) [reply crates/holochain_websocket/src/lib.rs:743]: Good spot, updated

**Review by jost-s — APPROVED** [review]

**Review by neonphog — APPROVED** [review]
:+1:

**Pull request merged**
